components/dashboard/src/data/git-providers/unified-repositories-search-query.ts

```diff
@@ -51,24 +33,52 @@ export const useUnifiedRepositorySearch = ({ searchString, excludeProjects = fal
     };
 };
 
-export const filterRepos = (searchString: string, suggestedRepos: SuggestedRepository[]) => {
-    let results = suggestedRepos;
+export function deduplicateAndFilterRepositories(
+    searchString: string,
+    excludeProjects = false,
+    suggestedRepos: SuggestedRepository[],
+): SuggestedRepository[] {
     const normalizedSearchString = searchString.trim();
-
-    if (normalizedSearchString.length > 1) {
-        results = suggestedRepos.filter((r) => {
-            return `${r.url}${r.projectName || ""}`.toLowerCase().includes(normalizedSearchString);
+    const collected = new Set<string>();
+    const results: SuggestedRepository[] = [];
+    const reposWithProject = new Set<string>();
+    if (!excludeProjects) {
+        suggestedRepos.forEach((r) => {
+            if (r.projectId) {
+                reposWithProject.add(r.url);
+            }
         });
-
-        if (results.length === 0) {
-            try {
-                // If the normalizedSearchString is a URL, and it's not present in the proposed results, "artificially" add it here.
-                new URL(normalizedSearchString);
-                results.push({ url: normalizedSearchString });
-            } catch {}
+    }
+    for (const repo of suggestedRepos) {
+        // filter out project entries if excludeProjects is true
+        if (repo.projectId && excludeProjects) {
+            continue;
+        }
+        // filter out project-less entries if an entry with a project exists
+        if (!repo.projectId && reposWithProject.has(repo.url)) {
+            continue;
+        }
+        // filter out entries that don't match the search string
+        if (!`${repo.url}${repo.projectName || ""}`.toLowerCase().includes(normalizedSearchString)) {
+            continue;
+        }
+        // filter out duplicates
+        const key = `${repo.url}:${repo.projectId || "no-project"}`;
+        if (collected.has(key)) {
+            continue;
         }
+        collected.add(key);
+        results.push(repo);
+    }
+
+    if (results.length === 0) {
+        try {
+            // If the normalizedSearchString is a URL, and it's not present in the proposed results, "artificially" add it here.
+            new URL(normalizedSearchString);
+            results.push({ url: normalizedSearchString });
+        } catch {}
     }
 
     // Limit what we show to 200 results
     return results.length > 200 ? results.slice(0, 200) : results;
```

Comment:
We could simplify this to just the slice. Slice doesn't mind the end index being higher than the array length.

```suggestion
    return results.slice(0, 200);
```

Reply:
Although maybe out of scope of these changes, I can also raise a follow-up 🙏.